src/Runtime/Distributions/DiscreteChar.cs

```diff
@@ -375,149 +234,91 @@ public bool IsPointMass
         [Construction("ProbabilityOutsideRanges", "GetRanges")]
         public static DiscreteChar Create(double probabilityOutsideRanges, IEnumerable<CharRange> ranges)
         {
-            var result = new DiscreteChar();
-            
             Argument.CheckIfNotNull(ranges, "ranges");
-            result.ranges = ranges.ToArray();
-            result.rangeCount = result.ranges.Length;
-
-            Array.Sort(result.ranges, (s1, s2) => Comparer<int>.Default.Compare(s1.StartInclusive, s2.StartInclusive));
-            CheckRanges(result.ranges, "ranges");
 
-            CheckUnnormalizedProbability(probabilityOutsideRanges, "probabilityOutsideRanges");
-            result.probabilityOutsideRanges = probabilityOutsideRanges;
-
-            result.Normalize();
-            result.MergeNeighboringRanges();
-
-            return result;
+            var builder = new StorageBuilder(probabilityOutsideRanges);
```

Comment:
This builder, do we really need to allocate a new object here or could we have that be a static class instead?

Reply:
It has state, so static means "thread local static" which is doable but meh.

I sort of hoped that .NET escape analysis will be smart enough to eliminate object allocation completely.

Probably I will help runtime a little and make StorageBuilder a (dreaded) mutable struct. With a huge comment that it's fine, because it's intended use case is to be created on stack and not passed around.